src/cache/mod.rs

```diff
@@ -417,10 +438,23 @@ impl<'a> ModuleCache<'a> {
 
     pub fn display_diagnostics(&self) {
         for diagnostic in &self.diagnostics {
-            eprintln!("{}", diagnostic.display());
+            let diagnostic = diagnostic.display(self);
+            eprintln!("{}", diagnostic);
         }
     }
 
+    pub fn get_contents(&mut self, path: &'a Path) -> Option<String> {
+        let contents = cached_read(&self.file_cache, path)?;
+        if !self.file_cache.contains_key(path) {
+            self.file_cache.insert(path, contents.clone());
+        };
+        Some(contents)
+    }
```

Comment:
```suggestion
    pub fn get_contents(&mut self, path: &'a Path) -> Option<&'a str> {
        let has_path = self.file_cache.contains_key(path);
        let contents = cached_read(&self.file_cache, path)?;
        
        if !has_path {
            let contents = contents.into_owned();
            self.file_cache.insert(path, contents);
        }
        
        self.file_cache.get(path).map(|s| s.as_str())
    }
```
A bit awkward to rewrite this to return a borrowed string when possible.